src/Uart/Uart.c

```diff
@@ -0,0 +1,441 @@
+/**@file
+ * This file is part of the Leon3 BSP for the Test Environment.
+ *
+ * @copyright 2022 N7 Space Sp. z o.o.
+ *
+ * Test Environment was developed under a programme of,
+ * and funded by, the European Space Agency (the "ESA").
+ *
+ *
+ * Licensed under the ESA Public License (ESA-PL) Permissive,
+ * Version 2.3 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://essr.esa.int/license/list
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/// \brief Uart hardware driver implementation.
+
+#include <stdbool.h>
+#include "Uart.h"
+#include "UartRegisters.h"
+#include "ByteFifo.h"
+#include "SystemConfig.h"
+#include <rtems.h>
+
+#define GPTIMER_ADDRESS_BASE 0x80000300U
+
+#ifdef RTEMS_SIS
+#define UART_INTERFACES_MAX 2
+#else
+#define UART_INTERFACES_MAX 6
+#endif
+
+static Uart_ErrorFlags errorFlags[UART_INTERFACES_MAX];
+
+static uint32_t sentBytesCounter[UART_INTERFACES_MAX];
+
+static rtems_interrupt_entry uartIrqEntry[UART_INTERFACES_MAX];
+
+static inline UartRegisters_t
+getAddressBase(Uart_Id id)
+{
+    switch(id) {
+        case Uart_Id_0:
+            return (UartRegisters_t)Uart0_address;
+        case Uart_Id_1:
+            return (UartRegisters_t)Uart1_address;
+        case Uart_Id_2:
+            return (UartRegisters_t)Uart2_address;
+        case Uart_Id_3:
+            return (UartRegisters_t)Uart3_address;
+        case Uart_Id_4:
+            return (UartRegisters_t)Uart4_address;
+        case Uart_Id_5:
+            return (UartRegisters_t)Uart5_address;
+        default:
+            return (UartRegisters_t)UartMax_address;
+    }
+}
+
+static inline Uart_BaudRate
+getBaudRate(const Uart* const uart)
+{
+    uint32_t freq = rtems_clock_get_ticks_per_second();
+    uint32_t baud = freq / (8 * ((uart->reg->clkscl << 8) + 1));
+    switch(baud) {
```

Comment:
Is it possible that baud won't have an exact value as the ones specified below? In that case, instead of returning the max baud rate value, shouldn't we return the closest value? Sometimes, it is not possible to set an exact baud rate from the standard values. If it is not possible then ok, we should not care about it.

Reply:
It should be tested on TSIM.